### PR TITLE
Do not parse arguments if log is disabled by priority

### PR DESCRIPTION
**Pull request by Pessimistress**
Skip calling `_getElapsedTime` etc. if the current priority is disabled.

### CODE REVIEW

**Comment by Pessimistress** [comment]
I see, I will open another diff to do that.
I do think this PR improves clarity when Probe is enabled, though.

**Comment by Pessimistress** [comment]
> Since the original methods are on the prototype, it should be possible to walk them and override them with noop fields on the object.

After some investigation I do not think it's going to work. `_shouldLog` returns true on `priority === 0` even if disabled. Which comes back to the point of this PR that we need a way to determine priority before parsing everything else.

File: src/lib/log.js [hunk 1]
```diff
@@ -382,57 +385,62 @@ in a later version. Use \`${newUsage}\` instead`);
...
       delta,
       total
     });
   }
 
-  // helper for _parseArguments
-  _normalizeArguments({priority, message, args = []}) {
-    let newOpts = null;
-
+  _normalizePriority(priority) {
```
Comment (ibgreen) [inline src/lib/log.js]: Maybe call it `_getPriorityFromFirstArgument`?

File: src/lib/log.js [hunk 2]
```diff
@@ -331,7 +331,10 @@ in a later version. Use \`${newUsage}\` instead`);
   // PRIVATE METHODS
 
   _shouldLog(priority) {
-    assert(Number.isFinite(priority), 'log priority must be a number');
+    priority = this._normalizePriority(priority);
+
+    assert(Number.isFinite(priority) && priority >= 0); // 'log priority must be a number'
```
Comment (ibgreen) [inline src/lib/log.js]: Should we do the assert inside the `_normalizePriority` method?

File: src/lib/log.js [hunk 3]
```diff
@@ -382,57 +385,62 @@ in a later version. Use \`${newUsage}\` instead`);
...
       break;
 
     default:
-      newOpts = {priority: 0, message, args};
+      newOpts = {message, args};
       break;
     }
 
-    assert(Number.isFinite(newOpts.priority)); // 'log priority must be a number'
+    newOpts.priority = this._normalizePriority(priority);
```
Comment (ibgreen) [inline src/lib/log.js]: Nit: I'd prefer calling `this._normalizePriority` at the top of this function and save the variable and then assign it here. Just to make it really clear that we are delegating to that function in all cases for the parsing of the first parameter.

File: src/lib/log.js [hunk 4]
```diff
@@ -382,57 +385,62 @@ in a later version. Use \`${newUsage}\` instead`);
...
     case 'function':
       if (message !== undefined) {
         args.unshift(message);
       }
-      newOpts = {priority: 0, message: priority, args};
+      newOpts = {message: priority, args};
       break;
 
     case 'object':
       const opts = priority;
```
Comment (ibgreen) [inline src/lib/log.js]: Hmm, how does this `opts` mix with the `opts` parameter? No linter complaints? Looks confusing, maybe could be written cleaner, perhaps renaming this helper variable to `argumentsObject` or similar?

File: src/lib/log.js [hunk 5]
```diff
@@ -382,57 +385,62 @@ in a later version. Use \`${newUsage}\` instead`);
...
   _parseArguments(options) {
-    const {priority, message, args = [], opts = {}} = options;
-    const normOpts = this._normalizeArguments({priority, message, args});
+    const normOpts = this._normalizeArguments(options);
 
     const {delta, total} = this._getElapsedTime();
 
     // original opts + normalized opts + opts arg + fixed up message + timings
-    return Object.assign(options, normOpts, opts, {
+    return Object.assign(options, normOpts, {
```
Comment (ibgreen) [inline src/lib/log.js]: Nit: Looks like comment needs to be updated?

**Pull request merged**